File: rust/src/2022/day6.rs
// vi: set shiftwidth=4 tabstop=4 expandtab:
use common::input::check_answer;
use common::input::get_answers;
use common::input::get_first_line_from_file;
use std::collections::HashSet;
use std::time::Instant;

const INPUT_FILEPATH: &str = "../resources/year2022_day6_input.txt";
const ANSWERS_FILEPATH: &str = "../resources/year2022_day6_answer.txt";

fn get_different_char(s: &str, winsize: usize) -> usize {
    for (i, w) in s
        .chars()
        .collect::<Vec<char>>()
        .windows(winsize)
        .enumerate()
    {
        let s = w.iter().copied().collect::<HashSet<char>>();
        if s.len() == winsize {
            return i + winsize;
        }
    }
    panic!("Unexpected");
}

fn part1(s: &str) -> usize {
    get_different_char(s, 4)
}

fn part2(s: &str) -> usize {
    get_different_char(s, 14)
}

fn main() {
    let before = Instant::now();
    let data = get_first_line_from_file(INPUT_FILEPATH);
    let (ans, ans2) = get_answers(ANSWERS_FILEPATH);
    let solved = true;
    let res = part1(&data);
    check_answer(&res.to_string(), ans, solved);
    let res2 = part2(&data);
    check_answer(&res2.to_string(), ans2, solved);
    println!("Elapsed time: {:.2?}", before.elapsed());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(part1("mjqjpqmgbljsphdztnvjfqwrcgsmlb"), 7);
        assert_eq!(part1("bvwbjplbgvbhsrlpgdmjqwftvncz"), 5);
        assert_eq!(part1("nppdvjthqldpwncqszvftbrmjlhg"), 6);
        assert_eq!(part1("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg"), 10);
        assert_eq!(part1("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"), 11);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2("mjqjpqmgbljsphdztnvjfqwrcgsmlb"), 19);
        assert_eq!(part2("bvwbjplbgvbhsrlpgdmjqwftvncz"), 23);
        assert_eq!(part2("nppdvjthqldpwncqszvftbrmjlhg"), 23);
        assert_eq!(part2("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg"), 29);
        assert_eq!(part2("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"), 26);
    }
}

File: python/2022/day6.py
# vi: set shiftwidth=4 tabstop=4 expandtab:
import datetime
import os


resource_dir = os.path.dirname(os.path.abspath(__file__)) + "/../../resources/"


def get_signal_from_file(file_path=resource_dir + "year2022_day6_input.txt"):
    with open(file_path) as f:
        for l in f:
            return l
    return None


def get_marker(signal, winsize):
    for i in range(len(signal)):
        if len(set(signal[i : i + winsize])) == winsize:
            return i + winsize
    return None


def run_tests():
    assert get_marker("mjqjpqmgbljsphdztnvjfqwrcgsmlb", winsize=4) == 7
    assert get_marker("bvwbjplbgvbhsrlpgdmjqwftvncz", winsize=4) == 5
    assert get_marker("nppdvjthqldpwncqszvftbrmjlhg", winsize=4) == 6
    assert get_marker("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", winsize=4) == 10
    assert get_marker("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", winsize=4) == 11
    assert get_marker("mjqjpqmgbljsphdztnvjfqwrcgsmlb", winsize=14) == 19
    assert get_marker("bvwbjplbgvbhsrlpgdmjqwftvncz", winsize=14) == 23
    assert get_marker("nppdvjthqldpwncqszvftbrmjlhg", winsize=14) == 23
    assert get_marker("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", winsize=14) == 29
    assert get_marker("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", winsize=14) == 26


def get_solutions():
    signal = get_signal_from_file()
    print(get_marker(signal, winsize=4) == 1850)
    print(get_marker(signal, winsize=14) == 2823)


if __name__ == "__main__":
    begin = datetime.datetime.now()
    run_tests()
    get_solutions()
    end = datetime.datetime.now()
    print(end - begin)
